monix/shared/src/main/scala/org/atnos/eff/addon/monix/TaskEffect.scala

```diff
@@ -0,0 +1,240 @@
+    TimedTask(_ => Task.async[A] { (_, cb) =>
+      callbackConsumer(tea => cb(tea.fold(Failure(_), Success(_))))
+    }, timeout).send
+
+}
+
+object TaskCreation extends TaskCreation
+
+trait TaskInterpretation extends TaskTypes {
+
+  def attempt[A](task: TimedTask[A]): TimedTask[Throwable Either A] = {
+    TimedTask[Throwable Either A](sexs => task.runNow(sexs).materialize.map(t => Either.cond(t.isSuccess, t.get, t.failed.get)))
+  }
+
+  def taskAttempt[R, A](e: Eff[R, A])(implicit task: TimedTask /= R): Eff[R, Throwable Either A] = {
+    e match {
+      case Pure(a, last) =>
+        Eff.pure[R, Throwable Either A](Either.right(a)).addLast(last)
+
+      case Impure(u, c, last) =>
+        task.extract(u) match {
+          case Some(tx) =>
+            val union = task.inject(attempt(tx))
+
+            Impure(union, Arrs.singleton { ex: (Throwable Either u.X) =>
+              ex match {
+                case Right(x) => taskAttempt(c(x))
+                case Left(t) => Eff.pure(Either.left(t))
+              }
+            }, last)
+
+          case None => Impure(u, Arrs.singleton((x: u.X) => taskAttempt(c(x))), last)
+        }
+
+      case ImpureAp(unions, continuation, last) =>
+        def materialize(u: Union[R, Any]): Union[R, Any] =
+          task.extract(u) match {
+            case Some(tx) => task.inject(attempt(tx).asInstanceOf[TimedTask[Any]])
+            case None => u
+          }
+
+        val materializedUnions =
+          Unions(materialize(unions.first), unions.rest.map(materialize))
+
+        val collected = unions.extract(task)
+        val continuation1 = Arrs.singleton[R, List[Any], Throwable Either A] { ls: List[Any] =>
+          val xors =
+            ls.zipWithIndex.collect { case (a, i) =>
+              if (collected.indices.contains(i)) a.asInstanceOf[Throwable Either Any]
+              else Either.right(a)
+            }.sequence
+
+          xors match {
+            case Left(t) => Eff.pure(Either.left(t))
+            case Right(anys) => taskAttempt(continuation(anys))
+          }
+        }
+
+        ImpureAp(materializedUnions, continuation1, last)
+    }
+  }
+
+  def memoize[A](key: AnyRef, cache: Cache, task: Task[A]): Task[A] =
+    Task.suspend {
+      cache.get[A](key).fold(task.map { r => cache.put(key, r); r })(Task.now)
+    }
+
+  implicit final def toTaskOps[R, A](e: Eff[R, A]): TaskOps[R, A] = new TaskOps[R, A](e)
+
+  /**
+    * Memoize tasks using a cache
+    *
+    * if this method is called with the same key the previous value will be returned
+    */
+  def taskMemo[R, A](key: AnyRef, cache: Cache, e: Eff[R, A])(implicit task: TimedTask /= R): Eff[R, A] = {
+    e match {
+      case Pure(a, last) =>
+        Pure(a, last)
+
+      case Impure(u, c, last) =>
+        task.extract(u) match {
+          case Some(tx) => Impure(task.inject(tx.copy(task = tx.task.andThen(memoize(key, cache, _)))), Arrs.singleton((x: u.X) => taskMemo(key, cache, c(x))), last)
```

Comment:
I am not sure this is the intended behaviour here. The idea is to cache the "computation to be made" and not its result. What I want is to avoid several threads asking the same data to an external service.